python/ray/serve/autoscaling_policy.py

```diff
@@ -0,0 +1,131 @@
+        # The number of replicas to be added when scaling up.
+        self.scale_up_num_replicas = config.get("scale_up_num_replicas", 2)
+        # The number of replicas to be removed when scaling down.
+        self.scale_down_num_replicas = config.get("scale_down_num_replicas", 1)
+        # The number of consecutive 'scale up' decisions that need to be made
+        # before the number of replicas is actually increased.
+        self.scale_up_consecutive_periods = config.get(
+            "scale_up_consecutive_periods", 2)
+        # The number of consecutive 'scale down' decisions that need to be made
+        # before the number of replicas is actually decreased.
+        self.scale_down_consecutive_periods = config.get(
+            "scale_down_consecutive_periods", 5)
+
+        # Keeps track of previous decisions. Each time the load is above
+        # 'scale_up_threshold', the counter is incremented and each time it is
+        # below 'scale_down_threshold', the counter is decremented. When the
+        # load is between the thresholds or a scaling decision is made, the
+        # counter is reset to 0.
+        self.decision_counter = 0
+
+    def scale(self, router_queue_lens, curr_replicas):
+        queue_lens = list(router_queue_lens.values())
+        if len(queue_lens) == 0:
+            return -1
+
+        avg_queue_len = sum(queue_lens) / len(queue_lens)
+
+        # Scale up.
+        if avg_queue_len > self.scale_up_threshold:
+            # Only actually scale the replicas if we've made this decision for
+            # 'scale_up_consecutive_periods' in a row.
+            if self.decision_counter >= self.scale_up_consecutive_periods - 1:
+                # TODO(edoakes): should we be resetting the counter here?
+                self.decision_counter = 0
+                decision = curr_replicas + self.scale_up_num_replicas
+                logger.info("Increasing number of replicas for backend '{}' "
+                            "from {} to {}".format(self.backend, curr_replicas,
+                                                   decision))
+            # Otherwise, just log this 'scale up' decision and do nothing.
+            else:
+                # If the previous decision was to scale down (the counter was
+                # negative), we reset it and then increment it (set to 1).
+                # Otherwise, just increment.
+                if self.decision_counter < 0:
+                    self.decision_counter = 1
+                else:
+                    self.decision_counter += 1
+                decision = -1
+
+        # Scale down.
+        elif avg_queue_len < self.scale_down_threshold and curr_replicas > 1:
+            # Only actually scale the replicas if we've made this decision for
+            # 'scale_down_consecutive_periods' in a row.
+            if (self.decision_counter <=
+                    -self.scale_down_consecutive_periods + 1):
```

Comment:
same here. `+1` is confusing and unnecessary if we do decision_counter book keeping first.